src/Plutarch/Test/QuickCheck.hs

```diff
@@ -0,0 +1,265 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImpredicativeTypes #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.Test.QuickCheck (
+    type IsPLam,
+    type IsLast,
+    type PTT,
+    type TestableFun,
+    type PLamArgs,
+    type PA,
+    type PB,
+    type PC,
+    FromPFunN (..),
+    fromPFun,
+    haskEquiv',
+    haskEquiv,
+    shrinkPLift,
+    arbitraryPLift,
+    PFun (..),
+    pattern PFn,
+    TestableTerm (..),
+    PArbitrary (..),
+    pconstantT,
+    pliftT,
+) where
+
+import Data.Kind (Type)
+import Generics.SOP (All, HPure (hcpure), NP ((:*)), Proxy (Proxy))
+import Plutarch (S, Term, (#), type (:-->))
+import Plutarch.Lift (PLift, PUnsafeLiftDecl (PLifted), plift)
+import Plutarch.Prelude (PBool, PInteger)
+import Plutarch.Show (PShow)
+import Plutarch.Test.QuickCheck.Function (
+    PFun (..),
+    applyPFun,
+    plamFinite,
+    plamTable,
+    pattern PFn,
+ )
+import Plutarch.Test.QuickCheck.Instances (
+    PArbitrary (..),
+    TestableTerm (..),
+    pconstantT,
+    pliftT,
+ )
+import Test.QuickCheck (
+    Arbitrary (..),
+    Gen,
+    Property,
+    Testable (property),
+    forAll,
+ )
+
+-- | @since 2.0.0
+data PTermType
+    = LastPFunction
+    | LastPTerm
+    | PFunction
+    | PTerm
+
+-- | @since 2.0.0
+type family PTT' isplam end :: PTermType where
```

Comment:
Kind signatures.

Reply:
@SeungheonOh what should the kinds be here?

Reply:
`Bool`, right?

Reply:
```hs
type family PTT' (isplam :: Bool) (end :: Bool) :: PTermType where
```

Sorry for being late, GH doesn't notified me this

Reply:
No worries, this was what I ended up going with.